### PR TITLE
fix[venom]: fix duplicate allocas

### PR DESCRIPTION
**Pull request by charles-cooper**
### Commit message

```
this commit fixes a bug in the ir_node_to_venom translator. previously,
`ir_node_to_venom` tried to detect unique allocas based on
heuristics. this commit removes the heuristics and fixes the issue
in the frontend by passing through a unique ID for each variable in
the metadata. this ID is also passed into the `alloca` and `palloca`
instructions to aid with debugging. note that this results in improved
code, presumably due to more allocas being able to be reified.

this commit makes a minor change to the `sqrt()`, builtin, which is to
use `z_var.as_ir_node()` instead of `z_var.pos`, since `.as_ir_node()`
correctly tags with the alloca metadata. to be maximally conservative,
we could branch, only using `z_var.as_ir_node()` if we are using
the venom pipeline, but the change should be correct for the legacy
pipeline as well anyways.
```
### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

**Comment by harkal** [comment]
`CurveStableSwapNG-0.4.1.vy` fails to compile with these changes

File: vyper/codegen/context.py [hunk 1]
```diff
@@ -15,13 +15,26 @@ class Constancy(enum.Enum):
     Constant = 1
 
 
+_alloca_id = 0
+
+
+def _generate_alloca_id():
+    # it is ok for this to be a global since the value does not affect generated code
```
Comment (cyberthirst) [inline vyper/codegen/context.py:22]: as per offline discussion this should be updated as it affects the code

**Comment by cyberthirst** [comment]
apart from the misleading comment lgtm. I agree with the commit msg in that it shouldn't be necessary to introduce venom-only path. checked only the legacy version.

**Comment by HodanPlodky** [comment]
There seems to be small regression on `VaultV3-0.4.0.vy` codesize other then that it LGTM

File: vyper/venom/function.py [hunk 2]
```diff
@@ -179,6 +179,25 @@ def chain_basic_blocks(self) -> None:
...
+            if bb is entry_bb:
+                continue
+
+            # "fast" way to strip allocas from each basic block
+            def is_alloca(inst):
+                return inst.opcode in ("alloca", "palloca")
+
+            bb.instructions.sort(key=is_alloca)
+            while len(bb.instructions) > 0 and is_alloca(bb.instructions[-1]):
+                entry_bb.insert_instruction(bb.instructions.pop())
```
Comment (harkal) [inline vyper/venom/function.py:197]:
```suggestion
            # Extract alloca instructions
            non_alloca_instructions = []
            for inst in bb.instructions:
                if inst.opcode in ("alloca", "palloca"):
                    entry_bb.insert_instruction(inst)
                else:
                    non_alloca_instructions.append(inst)

            # Replace original instructions with filtered list
            bb.instructions = non_alloca_instructions
```

**Review by harkal — APPROVED** [review]

File: vyper/venom/function.py (hunk 2 above)
Comment (charles-cooper) [reply vyper/venom/function.py:197]: this causes a slight regression in CurveStableSwapNG but a slight improvement in VaultV3.vy

**Pull request merged**